core/iwasm/libraries/lib-socket/test/tcp_ipv4.c

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (C) 2019 Intel Corporation.  All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ */
+#include <unistd.h>
+#include <string.h>
+#include <assert.h>
+#ifdef __wasi__
+#include <wasi_socket_ext.h>
+#endif
+#include <arpa/inet.h>
+#include <pthread.h>
+#define SERVER_MSG "Message from server."
+#define PORT 8989
+pthread_mutex_t mut;
+pthread_cond_t cond;
+int server_init_complete = 0;
+
+struct init_socket {
+    struct sockaddr_in addr;
+    int sock;
+};
+
+struct init_socket
+init_socket_addr()
+{
+    int sock = socket(AF_INET, SOCK_STREAM, 0);
+    assert(sock > 0);
+
+    struct sockaddr_in addr;
+    memset(&addr, 0, sizeof(addr));
+    addr.sin_family = AF_INET;
+    addr.sin_port = PORT;
+    addr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
+    return (struct init_socket){ addr, sock };
+}
+
+void *
+server()
+{
+    struct init_socket init_server_sock = init_socket_addr();
+    struct sockaddr_in server_addr = init_server_sock.addr, client_addr;
+    int server_sock = init_server_sock.sock, client_sock;
+    socklen_t addr_size;
+    char buffer[sizeof(SERVER_MSG) + 1];
+
+    int ret =
+        bind(server_sock, (struct sockaddr *)&server_addr, sizeof(server_addr));
+    assert(ret == 0);
+    pthread_mutex_lock(&mut);
+    server_init_complete = 1;
+    pthread_mutex_unlock(&mut);
+    pthread_cond_signal(&cond);
+
+    listen(server_sock, 1);
+    addr_size = sizeof(client_addr);
+    client_sock =
+        accept(server_sock, (struct sockaddr *)&client_addr, &addr_size);
+    assert(client_sock >= 0);
+
+    strcpy(buffer, SERVER_MSG);
+    sendto(client_sock, buffer, strlen(buffer), 0,
+           (struct sockaddr *)&client_addr, addr_size);
+
+    assert(close(client_sock) == 0);
+    return NULL;
+}
+
+void *
+client()
+{
+    pthread_mutex_lock(&mut);
+    while (server_init_complete == 0) {
+        pthread_cond_wait(&cond, &mut);
+    }
+
+    struct init_socket init_client_sock = init_socket_addr();
+    int sock = init_client_sock.sock;
+    pthread_mutex_unlock(&mut);
+    struct sockaddr_in addr = init_client_sock.addr;
+    socklen_t addr_size;
+    char buffer[sizeof(SERVER_MSG) + 1];
+    assert(connect(sock, (struct sockaddr *)&addr, sizeof(addr)) == 0);
+
+    recv(sock, buffer, sizeof(buffer), 0);
+    assert(strcmp(buffer, SERVER_MSG) == 0);
+    assert(close(sock) == 0);
+    return NULL;
+}
+
+int
+main(int argc, char **argv)
```

Comment:
There seems to be a lot of shared code between tcp_ipv4 and tcp_ipv6 test, we can potentially move the common code to separate file?

Reply:
done